### PR TITLE
Allow custom cluster types to read from and utilize eds_cluster_config

### LINKED ISSUE #15663: Custom clusters do not support endpoint discovery through EDS
*Title*: Custom clusters do not support endpoint discovery through EDS

*Description*:
Usage of eds_cluster_config is restricted to clusters that have type explicitly set to `EDS`. Envoy also uses this type for determining whether clusters are initialized during the primary or secondary phase, meaning it is not possible to write a custom cluster type that performs endpoint discovery via EDS.

Instead, I think the custom cluster type should specify whether it utilizes EDS for endpoint discovery, and if so, Envoy should treat it the same way as it does the native EDS cluster (i.e. the cluster is permitted to use `eds_cluster_config`, it is initialized during secondary phase if the config source is not file-based, etc.).

### PR DESCRIPTION
**Pull request by pianiststickman**
Commit Message: Allow custom cluster types to read from and utilize eds_cluster_config
Additional Description:
Usage of eds_cluster_config is currently restricted to clusters that have type explicitly set to EDS. Envoy also uses this type for determining whether clusters are initialized during the primary or secondary phase, meaning it is not possible to write a custom cluster type that performs endpoint discovery via EDS.

This CL introduces uses_eds_config to the CustomClusterType proto and changes the requirement for having eds_cluster_config set to include the case in which a custom cluster type has this flag set. Custom clusters that utilize the EDS config are omitted from the primary initialization phase, similar to clusters of type EDS.

Risk Level: low (small optional feature)
Testing: unit test
Docs Changes: inline on new field in cluster.proto
Release Notes: Adds a boolean field to CustomClusterType indicating that the custom cluster type uses EDS for endpoint discovery, and permits the custom cluster type to specify eds_cluster_config if that field is set.
Fixes #15663
API Considerations: the default value of uses_eds_config (false) results in no behaviour change (the cluster will be initialized during the primary phase; the cluster cannot specify eds_cluster_config).

Signed-off-by: Eugene Chan <eugenechan@google.com>

### CODE REVIEW

File: api/envoy/config/cluster/v3/cluster.proto [hunk 1]
```diff
@@ -175,6 +175,9 @@ message Cluster {
     // See the supported cluster for further documentation.
     // [#extension-category: envoy.clusters]
     google.protobuf.Any typed_config = 2;
+
+    // If true, this cluster type supports EDS updates and is configured using eds_cluster_config.
+    bool uses_eds_config = 3;
```
Comment (lizan) [inline api/envoy/config/cluster/v3/cluster.proto:180]: I don't think this should be exposed as part of API, instead the cluster implementation should indicate that to cluster manager at C++ level.

**Comment by snowp** [comment]
Is the purpose of this just to not be treated as a primary cluster? It definitely seems cleaner to have the cluster extensions themselves somehow announce that they should be treated as secondary clusters instead of requiring setting some additional config field.

File: api/envoy/config/cluster/v3/cluster.proto (hunk 1 above)
Comment (markdroth) [reply api/envoy/config/cluster/v3/cluster.proto:180]: Agreed, this should be part of the definition of the custom cluster type.  There's no reason the control plane should need to populate a field to tell the client to do this.

Comment (pianiststickman) [reply api/envoy/config/cluster/v3/cluster.proto:180]: As I see it, we run into a bit of a bootstrapping issue if we do that. At the time when we decide whether a cluster should be initialized during the primary or secondary phase, we haven't yet initialized the cluster such that the cluster implementation can indicate its preference. Indeed, we don't even know if the custom cluster type even exists yet; `loadCluster()` won't reject for an unknown type until we try to initialize the cluster.

We could expose the primary/secondary status of the cluster implementation in the factory implementation and pipe that through the ClusterManagerFactory (taking care to reject an implementation that doesn't exist), calling into it before the primary phase. I'm not sure I like the idea - it's added complexity to initialization (a custom cluster now gets "pre-initialized", just enough to tell whether to initialize during primary or secondary initialization), and also means that custom clusters get initialized differently than native clusters.

An alternative I'd probably favour is simply to allow custom clusters to set `eds_cluster_config` in addition to EDS clusters, and then use `has_eds_cluster_config()` to decide whether to initialize during primary/secondary phase. I think this is still technically config-driven, though it doesn't require any changes to the config API.

Thoughts appreciated.

**Comment by pianiststickman** [comment]
> Is the purpose of this just to not be treated as a primary cluster?

And to relax the restriction that only clusters that have the explicit type of `EDS` be allowed to set `eds_cluster_config`; otherwise, yes.

It does seem that at the time we begin the primary initialization phase we don't have the cluster implementations available to us, only the config. Thoughts would be appreciated on the review thread.

**Comment by markdroth** [comment]
As an alternative, why not have the config message for the custom cluster type include an `EdsClusterConfig` field of its own, rather than using the existing top-level `eds_cluster_config` field in the `Cluster` proto?  This would be consistent with the long-term goal of changing even the built-in cluster types to be extensions.

**Comment by maheshkurund** [comment]
@pianiststickman I would like to use EDS to fetch endpoints in custom cluster extension. I am interested to know how you used EDS in custom cluster.

**Pull request closed without merging**